### PR TITLE
[WIP] DirectCRonDirect

### PR DESCRIPTION
**Pull request by Vilin97**
Implemented `DirectCRonDirect` -- a spatial SSA that chooses the site with the composition-rejection search and chooses the reaction or hop via linear search (Direct method).

### CODE REVIEW

**Review by isaacsas — CHANGES_REQUESTED** [review]

File: src/spatial/DirectCRonDirect.jl [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+    # mapping from jump rate to group id
+    minexponent = exponent(minrate)
+
+    # use the largest power of two that is <= the passed in minrate
+    minrate = 2.0^minexponent
+    ratetogroup = rate -> priortogid(rate, minexponent)
+
+    # construct an empty initial priority table -- we'll reset this in init
+    rt = PriorityTable(ratetogroup, zeros(T, num_sites(spatial_system)), minrate, 2*minrate)
```
[the remark this answers is not in the document]
Comment (Vilin97) [reply src/spatial/DirectCRonDirect.jl:61]: Yes, I had forgotten to make these changes in the spatial DirectCR

File: src/spatial/DirectCRonDirect.jl [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    next_jump_time = typemax(typeof(end_time))
+    rx_rates = RxRates(num_sites(spatial_system), majumps)
+    hop_rates = HopRates(hopping_constants)
+    site_rates = zeros(typeof(end_time), num_sites(spatial_system))
+
+    DirectCRonDirectJumpAggregation(next_jump, next_jump_time, end_time, rx_rates, hop_rates, site_rates, save_positions, rng, spatial_system; num_specs = num_species, kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::DirectCRonDirectJumpAggregation, integrator, u, params, t)
```
[the remark this answers is not in the document]
Comment (Vilin97) [reply src/spatial/DirectCRonDirect.jl:85]: Yes. I can move it out to `initialize!` if you think it makes more sense.

File: src/spatial/DirectCRonDirect.jl [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+# calculate the next jump / jump time
+function generate_jumps!(p::DirectCRonDirectJumpAggregation, integrator, params, u, t)
+    @unpack rx_rates, hop_rates, rng, rt = p
+
+    p.next_jump_time  = t + randexp(rng) / rt.gsum
+    
+    if p.next_jump_time < p.end_time
+        site = sample(p.rt, p.site_rates, p.rng)
+        if rand(rng)*(total_site_rate(rx_rates, hop_rates, site)) < total_site_rx_rate(rx_rates, site)
```
[the remark this answers is not in the document]
Comment (Vilin97) [reply src/spatial/DirectCRonDirect.jl:99]: Yes, done. Good idea.

File: src/spatial/DirectCRonDirect.jl [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+"""
+function fill_rates_and_get_times!(aggregation::DirectCRonDirectJumpAggregation, u, t)
+    @unpack spatial_system, rx_rates, hop_rates, site_rates, rt, numspecies = aggregation
+
+    reset!(rx_rates)
+    reset!(hop_rates)
+    site_rates .= zero(typeof(t))
+
+    num_rxs = DiffEqJump.num_rxs(rx_rates)
+    num_sites = DiffEqJump.num_sites(spatial_system)
```
[the remark this answers is not in the document]
Comment (Vilin97) [reply src/spatial/DirectCRonDirect.jl]: For some reason my REPL sometimes complaints about it when there is no `DiffEqJump` at front

File: src/spatial/DirectCRDirect.jl [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+reevaluate all rates, recalculate tentative site firing times, and reinit the priority queue
+"""
+function fill_rates_and_get_times!(aggregation::DirectCRDirectJumpAggregation, u, t)
+    @unpack spatial_system, rx_rates, hop_rates, site_rates, rt, numspecies = aggregation
+
+    reset!(rx_rates)
+    reset!(hop_rates)
+    site_rates .= zero(typeof(t))
+
+    num_rxs = num_rxs(rx_rates)
```
Comment (isaacsas) [inline src/spatial/DirectCRDirect.jl:121]: Rename these local variables.

**Review by isaacsas — CHANGES_REQUESTED** [review]

File: src/spatial/DirectCRonDirect.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaacsas) [reply src/spatial/DirectCRonDirect.jl]: No it’s fine, just checking.

File: src/spatial/DirectCRonDirect.jl [hunk 6]
```diff
@@ -78,37 +78,27 @@ function aggregate(aggregator::DirectCRonDirect, starting_state, p, t, end_time,
...
-        else
-            species_to_diffuse, target_site = sample_hop_at_site(hop_rates, site, rng, p.spatial_system)
-            p.next_jump = SpatialJump(site, species_to_diffuse, target_site)
-        end
-    end    
+function generate_jumps!(p::DirectCRDirectJumpAggregation, integrator, params, u, t)
+    p.next_jump_time  = t + randexp(rng) / p.rt.gsum
+    p.next_jump_time >= p.end_time && return nothing
+    site = sample(p.rt, p.site_rates, p.rng)
+    p.next_jump = sample_jump_direct(p, site)
```
Comment (isaacsas) [inline src/spatial/DirectCRonDirect.jl]: Maybe call this something like `sample_jump_at_site_direct`?

Comment (isaacsas) [reply src/spatial/DirectCRonDirect.jl]: Actually the `direct` isn’t needed; other appraoches could use the same function and just dispatch (like RSSA).

**Review by isaacsas — APPROVED** [review]

**Pull request merged**